### PR TITLE
Speeds up the cloud-top-temperature CLI

### PR DESCRIPTION
**Pull request by MoseleyS**
Uses the existing mask to determine which columns need further processing. This thins the data more and more as the columns are ascended and results in a 3x speed-up for the MOGREPS-G global domain.

Testing:
 - [x] Ran tests and they passed OK

### CODE REVIEW

**Review by mspelman07 — DISMISSED** [review]
All looks good.

**Comment by Kat-90** [comment]
I have tested this change in the suite which succeeds with a wallclock resource of 560S.

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks for the update, @MoseleyS 👍 

I've made a few comments, mainly aimed at making the added code a bit clearer.

File: improver/psychrometric_calculations/cloud_top_temperature.py [hunk 1]
```diff
@@ -81,22 +80,36 @@ def _calculate_cct(self) -> ndarray:
...
         Temperature data are in Kelvin, Pressure data are in pascals, humidity data are in kg/kg.
+        A mask is used to keep track of which columns have already been solved, so that they
+        are not included in future iterations (x3 speed-up).
         """
         cct = np.ma.masked_array(self.t_at_ccl.data.copy())
         q_at_ccl = saturated_humidity(self.t_at_ccl.data, self.p_at_ccl.data)
-        ccl_with_mask = np.ma.where(True, self.t_at_ccl.data, False)
+        ccl_with_mask = np.ma.masked_array(
+            self.t_at_ccl.data, np.full_like(self.t_at_ccl.data, False, dtype=bool)
+        )
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/psychrometric_calculations/cloud_top_temperature.py:90]: I tried that as part of this work. The issue with it is that `ccl_with_mask.mask` becomes a scalar value of `False`, and doesn't interact with the stuff below. My code generates a multi-dimensional mask instead.

File: improver/psychrometric_calculations/cloud_top_temperature.py [hunk 2]
```diff
@@ -81,22 +80,36 @@ def _calculate_cct(self) -> ndarray:
...
+            t_2, _ = adjust_for_latent_heat(t_dry, q_at_ccl[~ccl_with_mask.mask], p)
             # Mask out points where parcel temperature, t_2, is less than atmosphere temperature, t,
             # but only after the parcel pressure, p, becomes lower than the cloud base pressure.
+            t_2_full = np.full_like(t.data, 99999)
+            t_2_full[~ccl_with_mask.mask] = t_2
+            t_loc_full = np.zeros_like(t.data)
+            t_loc_full[~ccl_with_mask.mask] = t_loc
             ccl_with_mask = np.ma.masked_where(
-                (t_2 < t_loc.data) & (p < self.p_at_ccl.data), ccl_with_mask,
+                (t_2_full < t_loc_full) & (p < self.p_at_ccl.data), ccl_with_mask,
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/psychrometric_calculations/cloud_top_temperature.py:107]: You might be right, but this looks much more complex to me, so I have left this bit as it is.

File: improver/psychrometric_calculations/cloud_top_temperature.py [hunk 3]
```diff
@@ -81,22 +80,36 @@ def _calculate_cct(self) -> ndarray:
...
+        ccl_with_mask = np.ma.masked_array(
+            self.t_at_ccl.data, np.full_like(self.t_at_ccl.data, False, dtype=bool)
+        )
         for t in self.temperature.slices_over("pressure"):
-            t_loc = copy.deepcopy(t)
+            t_loc = t.data[~ccl_with_mask.mask]
             (p,) = t.coord("pressure").points
-            t_dry = dry_adiabatic_temperature(self.t_at_ccl.data, self.p_at_ccl.data, p)
-            t_2, _ = adjust_for_latent_heat(t_dry, q_at_ccl, p)
+            t_dry = dry_adiabatic_temperature(
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/psychrometric_calculations/cloud_top_temperature.py:94]: Now named `pressure`

File: improver/psychrometric_calculations/cloud_top_temperature.py [hunk 4]
```diff
@@ -81,22 +80,36 @@ def _calculate_cct(self) -> ndarray:
...
         """
         cct = np.ma.masked_array(self.t_at_ccl.data.copy())
         q_at_ccl = saturated_humidity(self.t_at_ccl.data, self.p_at_ccl.data)
-        ccl_with_mask = np.ma.where(True, self.t_at_ccl.data, False)
+        ccl_with_mask = np.ma.masked_array(
+            self.t_at_ccl.data, np.full_like(self.t_at_ccl.data, False, dtype=bool)
+        )
         for t in self.temperature.slices_over("pressure"):
-            t_loc = copy.deepcopy(t)
+            t_loc = t.data[~ccl_with_mask.mask]
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/psychrometric_calculations/cloud_top_temperature.py]: I've gone with t_environment. I've also changed t_2 to t_parcel, and t_dry to t_dry_parcel.

[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/psychrometric_calculations/cloud_top_temperature.py]: Hmm - this seems to work, so I've put it in. Thanks.

**Comment by MoseleyS** [comment]
I'm passing this back to Gavin, but I haven't been able to reassure myself that these latest changes still deliver the speed-up I had achieved, mostly because I can't remember exactly what test I ran. I need to check this again before this can be merged.

**Review by gavinevans — APPROVED** [review]

**Comment by MoseleyS** [comment]
Finally worked out what I'd got wrong with my test - all is well. This can now be merged.

**Pull request merged**